Add support for Perl(PCRE) named and unnamed group capturing order

CIAvash opened the pull request:
In other words maintain the order of capture groups. With the `MaintainCaptureOrder` regexp option.

I also added inline option `o`. It's useful if you only have access to the pattern, but not the regex. But it only is useful if used at the start of the pattern, I couldn't find a way to prevent it from being used elsewhere.

I've never liked nor have been good with bitwise operations. So I don't know if I should've picked another number.

CIAvash committed:
Add support for Perl(PCRE) named and unnamed group capturing order.

In other words maintain the order of capture groups.

CIAvash committed:
Check the length of `p.capnamelist` when assigning slots

CIAvash committed:
Add test for when no capture group is used with inline option

dlclark commented on README.md:
```diff
@@ -74,6 +74,7 @@ The internals of `regexp2` always operate on `[]rune` so `Index` and `Length` da
 | named back reference `\k'name'` | no | yes |
 | named ascii character class `[[:foo:]]`| yes | no (yes in RE2 compat mode) |
 | conditionals `(?(expr)yes\|no)` | no | yes |
+| PCRE capture group order | no | no (yes in MaintainCaptureOrder mode) |
```
We should document the existence of the modifier and the specific behavior of the mode in the readme.

dlclark commented on syntax/parser.go:
```diff
@@ -43,6 +44,8 @@ func optionFromCode(ch rune) RegexOptions {
 		return Debug
 	case 'e', 'E':
 		return ECMAScript
+	case 'o', 'O':
```
Do we really want patterns to be able to change the output of the `Groups()` function?  It seems like the programmer creating the `regexp` struct would be expecting certain behavior and allowing patterns to change it could be surprising (especially if the patterns are dynamic in any way).

Also, the `o` modifier has a meaning in PCRE apparently so if we decide to keep the modifier concept then we should use something besides `o`.  https://stackoverflow.com/questions/22937618/reference-what-does-this-regex-mean/22944075#22944075

CIAvash replied:
I think you are right, I removed the inline option.

dlclark commented on syntax/parser.go:
```diff
@@ -214,80 +218,99 @@ func (p *parser) noteCaptureName(name string, pos int) {
+		// Prepend `0` to capnamelist if it's not set (MaintainCaptureOrder was enabled inline)
+		if len(p.capnamelist) == 0 || p.capnamelist[0] != `0` {
+			p.capnamelist = append([]string{fmt.Sprint(0)}, p.capnamelist...)
+		}
```
can this just return and not use the else?  should clean up the diff quite a bit.

dlclark commented on syntax/parser.go:
```diff
@@ -386,7 +417,11 @@ func (p *parser) countCaptures() error {
 					if !p.useOptionN() && !p.ignoreNextParen {
-						p.noteCaptureSlot(p.consumeAutocap(), pos)
+						if p.useMaintainCaptureOrder() {
+							p.noteCaptureName(fmt.Sprint(p.autocap), pos)
```
why not `consumeAutocap()` ?

CIAvash replied:
Because `consumeAutocap` is called inside `noteCaptureName` when `MaintainCaptureOrder` is enabled.

Does this need a comment somewhere?

dlclark commented on syntax/parser.go:
```diff
@@ -921,6 +956,10 @@ func (p *parser) scanGroupOpen() (*regexNode, error) {
 						return nil, p.getErr(ErrInvalidGroupName)
 					}
+
+					if capnum != -1 && p.useMaintainCaptureOrder() {
```
a comment here would be nice, why do consume the Autocap?  I assume it's because we've just successfully setup a named capture group so we need to increment our cap numbers to maintain the order in this mode

dlclark commented on regexp_MaintainCaptureOrder_test.go:
```diff
@@ -0,0 +1,394 @@
+)
+
+func TestMaintainCaptureOrder_Basic(t *testing.T) {
+	r, err := Compile("(?<first>this).+?(testing).+?(?<last>stuff)", MaintainCaptureOrder)
```
can we get a sister test where `MaintainCaptureOrder` isn't set?  That'd let us confirm the old behavior of `Groups()` is unaffected.

dlclark reviewed (commented):
On first pass I have a few questions, it'd be great to get your perspective on them.

CIAvash committed:
Applied requested changes

- Document MaintainCaptureOrder option
- Use return in `assignNameSlots` and remove else
- Add test with MaintainCaptureOrder not provided
- Change the MaintainCaptureOrder value to `0x0400`
- Remove the `o` inline option
- Add comment to explain why `autocap` is consumed

yywing commented:
Is there anyway get named group？

like `(1)(?<aa>2)`

I want get `aa` only.

CIAvash commented:
@yywing Don't know why you asked that question here, as it's not relevant to the pull request. But you can do ``match.GroupByName(`aa`)``.

CIAvash committed:
Use the `String` method instead of converting runes to string
